spec.bs

```diff
@@ -0,0 +1,1484 @@
+* An { <code>S</code> <code class="vocab">tree:validatedBy</code> <code>Sh</code> } arc indicates that <code>R</code> must have exactly one node which conforms to shape <code>Sh</code>.
+* Any { <code>S</code> <code class="vocab">tree:contains</code> <code>S2</code> } arcs indicate there is a nested [=shape tree=] and corresponding nested resource <code>R2</code>.
+
+A { <code class="vocab">tree:hasShapeTreeDecoratorIndex</code> <code>Si</code> } arc indicates the location of an index of SKOS hierarchies that describe <code>ST</code>.
+
+<figure id="shapetree-shex">
+  <figcaption>ShEx validation of a Shape Tree</figcaption>
+	<pre highlight="turtle">
+  prefix tree: &lt;http://www.w3.org/ns/shapetree#&gt;.
+  prefix rdfs: &lt;http://www.w3.org/2000/01/rdf-schema#&gt;.
+  prefix xsd: &lt;http://www.w3.org/2001/XMLSchema#&gt;.
+
+	&lt;#ShapeTree&gt; {
+    a tree:ShapeTree ;
+	  (
+	    tree:expectsType [tree:ShapeTreeContainer] ;
+	    tree:contains @&lt;#ShapeTree&gt; +
+	  ) ;
+    (rdfs:label xsd:string | tree:matchesUriTemplate xsd:string) ;
+	  tree:references @&lt;#ReferencedShapeTree&gt; * ;
+	  tree:validatedBy IRI ? ;
+    tree:contains IRI ? ;
+    tree:supports IRI ? ;
+	}
+
+	&lt;#ReferencedShapeTree&gt; {
+	  tree:hasShapeTree IRI ;
+	  tree:traverseViaShapePath xsd:string
+	}
+	</pre>
+</figure>
+
+Ecosystems {#ecosystems}
+=====================
+[=Shape trees=] are meant to be used in context of a broader [=ecosystem=].  [=Ecosystems=] have the ability to react to (and in some narrow cases influence) [=shape tree=] operations.
+
+[=Ecosystems=] are helpful for use cases such as indexing the location of [=planted shape trees=].
+
+[=Shape tree=] client libraries should be initialized with a reference to the the target [=ecosystem=] it should operate within.
+
+For purposes of this specification, an [=ecosystem=] should be considered a software interface, expecting the below operations to be implemented.
+
+## Ecosystem Operations ## {#ecosystem-operations}
+
+### Get Existing Shape Tree ### {#ecosystem-getexistingshapetree}
+<code><b>getExistingShapeTreeFromContainer(context:ShapeTreeContext, parentContainer:URI, shapeTreesToPlant:ShapeTree[], resourceName:String) : ShapeTreeLocator</b></code>
+
+This operation allows the [=ecosystem=] to provide a previously [=planted shape tree=] location.
+
+### Before Plant Shape Tree ### {#ecosystem-beforeplantshapetree}
+<code><b>beforePlantShapeTree</b>(context:ShapeTreeContext, targetContainer:URI, bodyGraph:Graph, shapeTreesToPlant:ShapeTree[], linkHeaders:Map) : Graph</code>
+
+This operations allows the [=ecosystem=] to modify the the [=body graph=] of a [Plant Shape Tree](#plant-shapetree) operation request.
+
+### Index Shape Tree ### {#ecosystem-indexshapetree}
```

Comment:
As of now there really isn't a notion of updating a planted shape tree.  It can be unplanted, but there aren't currently any things that can 'change'.  For instance data that is changed, while there is nothing metadata-related that changes, it would not hurt to add an ecosystem hook to be notified of changed data.